stm32 sample application to test low power modes

FRASTM opened the pull request:
This PR configures the Power Management policy for the stm32 soc series
and gives a sample application to enter/exit the low power mode.

The power residency is required to set the minimum timeout for entering sleep mode.
The sample application is demonstrating the power saving mode on the k_sleep with wakeup by the user button or the given timeout.
The CONFIG_SYS_POWER_MANAGEMENT is required to enable the low power modes.
In case the timeout is changed to 'FOREVER' only the user button can wakeup the kernel.
The PR #25978 is required then.
The different values of each SYS_PM_MIN_RESIDENCY is set for evaluation purpose significantly higher than the corresponding "time to wake Up"

The user button in this application corresponds to the WAKEUP PIN2 (= GPIO PC13)

The Deep Sleep mode is not supported by this PR.

Signed-off-by: Francois Ramu francois.ramu@st.com

FRASTM commented:
change type of var

FRASTM commented:
rebase

FRASTM commented:
rebase

erwango commented on subsys/power/policy/Kconfig:
```diff
@@ -7,8 +7,9 @@ choice
 config SYS_PM_POLICY_RESIDENCY
 	bool "PM Policy based on CPU residency"
-	select SYS_PM_POLICY_RESIDENCY_DEFAULT if !SOC_SERIES_CC13X2_CC26X2
+	select SYS_PM_POLICY_RESIDENCY_DEFAULT if !SOC_SERIES_CC13X2_CC26X2 && !SOC_SERIES_STM32L4X
```
```suggestion
	select SYS_PM_POLICY_RESIDENCY_DEFAULT if !SOC_SERIES_CC13X2_CC26X2 && !SOC_FAMILY_STM32
```

FRASTM replied:
ok, to set to all stm32xx

erwango commented on subsys/power/policy/Kconfig:
```diff
@@ -7,8 +7,9 @@ choice
-	select SYS_PM_POLICY_RESIDENCY_DEFAULT if !SOC_SERIES_CC13X2_CC26X2
+	select SYS_PM_POLICY_RESIDENCY_DEFAULT if !SOC_SERIES_CC13X2_CC26X2 && !SOC_SERIES_STM32L4X
 	select SYS_PM_POLICY_RESIDENCY_CC13X2_CC26X2 if SOC_SERIES_CC13X2_CC26X2
+	select SYS_PM_POLICY_RESIDENCY_STM32 if SOC_SERIES_STM32L4X
```
```suggestion
	select SYS_PM_POLICY_RESIDENCY_STM32 if SOC_FAMILY_STM32
```

erwango commented on samples/subsys/power/stm32_pm/src/main.c:
```diff
@@ -0,0 +1,103 @@
+/* in msec */
+#define WAIT_TIME 500
+#define SLEEP_TIME 4001
+//#define SLEEP_TIME K_TICKS_FOREVER
```
Please remove this line

FRASTM replied:
commented

erwango commented on samples/subsys/power/stm32_pm/src/main.c:
```diff
@@ -0,0 +1,103 @@
+
+#define DEMO_DESCRIPTION	\
+	"Demo Description\n"	\
+	"Application is demonstrating the LPTIMER use as systick.\n"\
```
```suggestion
	"Application is demonstrating the LPTIMER used as systick.\n"\
```

FRASTM replied:
fixed

erwango commented on subsys/power/policy/policy_residency_stm32.c:
```diff
@@ -0,0 +1,73 @@
+#include <kernel.h>
+#include "pm_policy.h"
+
+#include <stm32l4xx_ll_lptim.h>
```
Is that still required ?
If yes, this won't work with other series than L4

FRASTM replied:
also removing references to LPTIM clock

erwango commented on samples/subsys/power/stm32_pm/sample.yaml:
```diff
@@ -0,0 +1,13 @@
+  name: Low Power Timer Sample for nucleo board
+tests:
+  sample.power.pm.stm32_lptim.lsi:
+    platform_whitelist: nucleo_l4r5zi
```
can we try to remove withelist and filter on LTPIM support ? 
This would enable testing on more platforms

FRASTM replied:
no, removed

erwango commented on samples/subsys/power/stm32_pm/src/main.c:
```diff
@@ -0,0 +1,103 @@
+void button_pressed(struct device *gpiob, struct gpio_callback *cb,
+		    uint32_t pins)
+{
+	LL_PWR_ClearFlag_WU();
```
ARG_UNUSED(gpiob);
	ARG_UNUSED(pins);
	ARG_UNUSED(cb);

FRASTM replied:
ok

FRASTM reviewed (commented):
actually, at this time, there is no reason to change the policy for stm32,
until more specific action is required
So keeping the generic SYS_PM_POLICY_RESIDENCY

FRASTM commented:
1. do not change the SYS_PM_POLICY_RESIDENCY

2. configure nucleo board  for test sequence

ABOSTM commented on samples/subsys/power/stm32_pm/sample.yaml:
```diff
@@ -0,0 +1,17 @@
+    filter: dt_compat_enabled_with_alias("gpio-leds", "led0") and
+            dt_compat_enabled_with_alias("gpio-leds", "led1") and
+            dt_compat_enabled("st,stm32-lptim")
+  sample.power.pm.stm32_lptim.lse:
+    build_only: true
+    depends_on: lptim
+    filter: dt_compat_enabled_with_alias("gpio-leds", "led0") and
+            dt_compat_enabled_with_alias("gpio-leds", "led1") and
+            dt_compat_enabled("st,stm32-lptim")
+    extra_configs:
+      - CONFIG_STM32_LPTIM_CLOCK_LSE=y
```
Is it possible to exclude boards that don't have a LSE ?

FRASTM replied:
changed to have only 1 led : off = low power, on=running

erwango replied:
> Is it possible to exclude boards that don't have a LSE ?

@ABOSTM, it is not possible I think . But we're filtering on boards that declare `lptim` support and this should encompass LSE support.

ABOSTM commented on samples/subsys/power/stm32_pm/src/main.c:
```diff
@@ -0,0 +1,103 @@
+
+/* in msec */
+#define WAIT_TIME 500
+#define SLEEP_TIME 4001
```
Can you add a comment about those while ? why 4001 ?

FRASTM replied:
changed to set any timeout value

ABOSTM commented on samples/subsys/power/stm32_pm/src/main.c:
```diff
@@ -0,0 +1,103 @@
+	"Application is demonstrating the LPTIMER used as systick.\n"\
+	"The system goes to sleep mode on k_sleep() and wakes-up\n" \
+	"(red led on) by the lptim timeout expiration\n" \
+	"or when pressing the user push button (WUP2 is PC13)\n\n"
```
> WUP2 is PC13

Is this valid for all STM32 series ?

FRASTM replied:
most of the stm32 boards have a push button, if the soc has lptim

ABOSTM commented on samples/subsys/power/stm32_pm/src/main.c:
```diff
@@ -0,0 +1,103 @@
+{
+	LL_PWR_ClearFlag_WU();
+	/* signal wakeUp from pin */
+	gpio_pin_set(gpio_led2, PIN2, (int)LED_ON);
```
> 	gpio_pin_set(gpio_led2, PIN2, (int)LED_ON);

Is PIN2 valid for all STM32 serries ?

ABOSTM commented on samples/subsys/power/stm32_pm/src/main.c:
```diff
@@ -0,0 +1,103 @@
+
+	while (1) {
+		/* Set pin to HIGH/LOW */
+		gpio_pin_set(gpio_led, PIN0, LED_ON);
```
> gpio_pin_set(gpio_led, PIN0, LED_ON);

Is PIN0 valid for all STM32 serries ?

FRASTM commented on samples/subsys/power/stm32_pm/src/main.c:
```diff
@@ -0,0 +1,103 @@
+	gpio_bn = device_get_binding(SW);
+
+
+	if (IS_ENABLED(CONFIG_SYS_POWER_MANAGEMENT)) {
```
condition removed

pabigot replied:
It seems to still be there.

pabigot commented on samples/subsys/power/stm32_pm/sample.yaml:
```diff
@@ -0,0 +1,17 @@
+sample:
+  name: Low Power Timer Sample for nucleo board
+tests:
+  sample.power.pm.stm32_lptim.lsi:
```
This might be better placed at `samples/boards/stm32/lptim` or otherwise under `samples/boards/stm32`, as it's demonstrating something related to how a particular family of boards works, and several other systems (TI, Nordic, mec15xxevb) have placed power management demonstrations there.

FRASTM replied:
moving to a new `samples/boards/stm32/pm_lptim`

pabigot commented on samples/subsys/power/stm32_pm/src/main.c:
```diff
@@ -0,0 +1,89 @@
+
+/* setting to K_TICKS_FOREVER willl activate the deepsleep mode */
+
+#define DEMO_DESCRIPTION	\
```
This is nice, but there should be a `README.rst` that documents this sample and is integrated into the built documentation, as there is for most other samples.

FRASTM replied:
done

pabigot commented on samples/subsys/power/stm32_pm/src/main.c:
```diff
@@ -0,0 +1,89 @@
+void button_pressed(struct device *gpiob, struct gpio_callback *cb,
+		    uint32_t pins)
+{
+		ARG_UNUSED(gpiob);
```
Indentation here is wrong (too deep).  Try uncrustify.

pabigot commented on samples/subsys/power/stm32_pm/src/sample.h:
```diff
@@ -0,0 +1,55 @@
+
+#include <zephyr.h>
+
+#include <string.h>
```
Nothing in this file is referenced except from `main.c`.  IMO its contents should be put into `main.c`.

FRASTM replied:
ok, removed the sample.h

pabigot commented on boards/arm/nucleo_l4r5zi/nucleo_l4r5zi.yaml:
```diff
@@ -17,5 +17,6 @@ supported:
   - usb device
   - nvs
   - counter
+  - lptim
```
Please also add this to all other boards that would support this feature.

FRASTM replied:
For the moment only this one was actually tested. It would be easy though to include for any board where the soc supports the feature.

erwango commented on samples/subsys/power/stm32_pm/src/main.c:
```diff
@@ -0,0 +1,124 @@
+#include <drivers/gpio.h>
+#include <sys/printk.h>
+
+struct device *gpio_bn, *gpio_led, *gpio_led2;
```
```suggestion
struct device *gpio_bn, *gpio_led;
```

FRASTM replied:
one led only

erwango commented on samples/subsys/power/stm32_pm/sample.yaml:
```diff
@@ -0,0 +1,17 @@
+  sample.power.pm.stm32_lptim.lsi:
+    build_only: true
+    depends_on: lptim
+    filter: dt_compat_enabled_with_alias("gpio-leds", "led0") and
+            dt_compat_enabled_with_alias("gpio-leds", "led1") and
+            dt_compat_enabled("st,stm32-lptim")
```
```suggestion
    filter: dt_compat_enabled_with_alias("gpio-leds", "led0") and
            dt_compat_enabled("st,stm32-lptim")
```

FRASTM replied:
no more led1

erwango commented on samples/subsys/power/stm32_pm/sample.yaml:
```diff
@@ -0,0 +1,17 @@
+  sample.power.pm.stm32_lptim.lse:
+    build_only: true
+    depends_on: lptim
+    filter: dt_compat_enabled_with_alias("gpio-leds", "led0") and
+            dt_compat_enabled_with_alias("gpio-leds", "led1") and
+            dt_compat_enabled("st,stm32-lptim")
```
```suggestion
    filter: dt_compat_enabled_with_alias("gpio-leds", "led0") and
            dt_compat_enabled("st,stm32-lptim")
```

FRASTM replied:
no more led1

FRASTM commented:
Include: a README file

erwango reviewed (approved):
Approved on my side.

ABOSTM commented on samples/subsys/power/stm32_pm/src/main.c:
```diff
@@ -0,0 +1,123 @@
+#else
+/* A build error here means your board isn't set up to blink an LED. */
+#error "Unsupported board: led0 devicetree alias is not defined"
+#define LED0	""
+#define PIN0	0
```
Those lines seems useless as there is #error above

FRASTM replied:
yes, removed

ABOSTM commented on samples/subsys/power/stm32_pm/src/main.c:
```diff
@@ -0,0 +1,123 @@
+void button_pressed(struct device *gpiob, struct gpio_callback *cb,
+		uint32_t pins)
+{
+	ARG_UNUSED(gpiob);
+	ARG_UNUSED(pins);
+	ARG_UNUSED(cb);
```
If parameter are useless why not removing them ?
Note: gpiob could be replaced by gpio to avoid confusion with gpio port B

FRASTM replied:
this is the format of the `gpio_init_callback`  model which is inherited from `samples/basic/button/src
`

FRASTM replied:
rename `gpio_bn` for button

ABOSTM commented on samples/subsys/power/stm32_pm/src/main.c:
```diff
@@ -0,0 +1,123 @@
+
+/* in msec */
+#define SLEEP_TIME 3000
+/* setting to K_TICKS_FOREVER willl activate the deepsleep mode */
```
This comment could be moved before the #define

FRASTM replied:
ok

pabigot reviewed (commented):
AFAICT this PR is preparing a sample that will work when power management is supported.  I'm not clear whether it's supposed to have an effect on power consumption now, so haven't tried to measure it.  I've played with this on nucleo_l476rg.  It does indeed turn off the LED for 3 s when the button is pushed.

For it to be a good example for verifying power consumption it needs some changes.

First, how exactly do you measure power on the board?  I ask because if an LED is lit AFAICT the power draw of the LED obscure the impact of putting the MCU into a low power state.  If ST development boards usually have a way to measure only MCU power draw, then a mention of where to find that information would be helpful.

Second, the sample configures the residency mode with three durations to enable all three sleep modes.  Is there a functional different between them?  If so, then it would make sense to cycle through sleeps that trigger each mode in turn.  If not, please only demonstrate the supported modes so users aren't misled by the suggestion that there's a difference.

FRASTM commented:
rebase and move example to a new **`samples/subsys/power/stm32_pm`**

erwango commented on samples/boards/stm32/pm_lptim/sample.yaml:
```diff
@@ -0,0 +1,15 @@
+sample:
+  name: Power Management LP Timer
+tests:
+  sample.power.pm.stm32_lptim.lsi:
```
```suggestion
  sample.boards.stm32.pm_lptim.lsi:
```

erwango commented on samples/boards/stm32/pm_lptim/sample.yaml:
```diff
@@ -0,0 +1,15 @@
+    depends_on: lptim
+    filter: dt_compat_enabled_with_alias("gpio-leds", "led0") and
+            dt_compat_enabled("st,stm32-lptim")
+  sample.power.pm.stm32_lptim.lse:
```
```suggestion
  sample.boards.stm32.pm_lptim.lse:
```

ABOSTM commented on samples/boards/stm32/pm_lptim/prj.conf:
```diff
@@ -0,0 +1,12 @@
+CONFIG_TIMESLICING=n
+CONFIG_DEBUG=y
```
Is it necessary to keep CONFIG_DEBUG ?

FRASTM replied:
no, removed

pabigot commented on samples/boards/stm32/pm_lptim/README.rst:
```diff
@@ -0,0 +1,44 @@
+Building and Running
+********************
+
+Build and flash sample as follows, changing ``reel_board`` for your board:
```
Since this is an STM32-specific test you should use a known-working STM32 board rather than `reel_board` in the example.

FRASTM replied:
changed to 'nucleo_l4r5zi'

ABOSTM commented on samples/boards/stm32/pm_lptim/sample.yaml:
```diff
@@ -0,0 +1,15 @@
+tests:
+  sample.boards.stm32.pm_lptim.lsi:
+    build_only: true
+    depends_on: lptim
```
Is it possible to make Power Management on all soc supporting lptim?
Or should we add a dependance on power management ?

FRASTM replied:
No, at this step, the LPTIM is used for  counting time (kernel ticks) _only_ in case the soc is activating  the power management

ABOSTM commented on samples/boards/stm32/pm_lptim/sample.yaml:
```diff
@@ -0,0 +1,15 @@
+    filter: dt_compat_enabled_with_alias("gpio-leds", "led0") and
+            dt_compat_enabled("st,stm32-lptim")
+  sample.boards.stm32.pm_lptim.lse:
+    build_only: true
```
Should we consider changing the test os that it could be not only build but also executed. and thus be run on test bench ?

erwango replied:
@FRASTM ?

FRASTM replied:
When executing the test case automatically, it happens that console is no more responding
That's what the https://github.com/zephyrproject-rtos/zephyr/issues/31284 reports.
Fixing this is prefered before adding the test case to the bench.

pabigot commented on boards/arm/nucleo_l4r5zi/nucleo_l4r5zi.yaml:
```diff
@@ -17,5 +17,6 @@ supported:
   - usb device
   - nvs
   - counter
+  - lptim
```
Works on nucleo_l476rg and nucleo_wb55rg.

pabigot commented on samples/boards/stm32/pm_lptim/README.rst:
```diff
@@ -0,0 +1,44 @@
+
+After flashing, the LED is on, waiting for entering the low Power mode.
+Press the User button to enter the low power mode. The led is off.
+On timeout, exiting the low power mode, the led is On again.
```
Tell the user how long to expect to wait, e.g. "After three seconds the timeout wake the system and turn the led on again."

pabigot commented on samples/boards/stm32/pm_lptim/src/main.c:
```diff
@@ -0,0 +1,119 @@
+		}
+
+		gpio_pin_set(gpio_led, PIN0, 1);
+		k_busy_wait(10000);
```
Is the busy wait necessary?  Seems simpler to just let the thing loop.  It won't hurt to set the GPIO active repeatedly, will it?

FRASTM replied:
right

pabigot commented on samples/boards/stm32/pm_lptim/src/main.c:
```diff
@@ -0,0 +1,119 @@
+	ARG_UNUSED(pins);
+	ARG_UNUSED(cb);
+
+	/* Set pin to HIGH/LOW */
```
The comment is confusing.  How about:

    /* Turn LED off */

if this needs a comment at all.

FRASTM committed:
samples: boards: stm32 new example for power management

This sample uses the lptim when calling the k_sleep
for the STM32xx from STMicroelectronics.
Using new DT for button and Led,
Button is ready to exit from low power state.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

pabigot commented on samples/boards/stm32/pm_lptim/src/main.c:
```diff
@@ -0,0 +1,119 @@
+#if DT_NODE_HAS_STATUS(SW0_NODE, okay)
+#define SW	DT_GPIO_LABEL(SW0_NODE, gpios)
+#define PIN	DT_GPIO_PIN(SW0_NODE, gpios)
+#if DT_PHA_HAS_CELL(SW0_NODE, gpios, flags)
```
You don't need to make this conditional now that #27076 has been merged.

FRASTM replied:
this part is similar to the existing `samples/basic/button/src/main.c`

pabigot replied:
No, it's similar to `button` before the sample got fixed up.

FRASTM replied:
right, removed

pabigot reviewed (changes requested):
Referencing a relevant board in the readme is probably the only blocker, but the others are worth considering.

erwango commented:
CI failure should be fixed by https://github.com/zephyrproject-rtos/zephyr/pull/28065

FRASTM committed:
boards: arm: several nucleo boards activates the LPTIM feature

This will activate the LPTIM for the test sequence
once the power management is enabled

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM reviewed (commented):
rebased

pabigot commented:
@frastm You'll need to "reopen and comment" if you want it re-tested.  Please don't open a new PR.

FRASTM commented:
reopening

FRASTM commented:
@galak can you please reopen this PR (option is not available for us). Thank you

pabigot commented:
Ah.  I see; github prevents re-opening once the branch has been force-pushed.  I got bit by that before myself.

If you can reconstruct the commit that was on this branch at the time you force-pushed yesterday you can force-push that up, and it *should* be possible to re-open, then push the updated content.

If you can't push the old content up, or doing so doesn't unlock the "Reopen and comment" button, try deleting the github branch, then recreating it with the new content.  That still might not let you re-open this PR, but you could then open a new one.

FRASTM commented:
re-openning and push after rebase

erwango commented on samples/boards/stm32/pm_lptim/src/main.c:
```diff
@@ -0,0 +1,109 @@
+
+/* in msec */
+/* setting to K_TICKS_FOREVER willl activate the deepsleep mode */
+#define SLEEP_TIME 11000
```
11 sec is a bit long (I sometime worry the use case is broken). Maybe 3 sec would be enough ?

FRASTM replied:
yes, reduced (but still modifiable by the user)

FRASTM commented on samples/boards/stm32/pm_lptim/src/main.c:
```diff
@@ -0,0 +1,127 @@
+			GPIO_OUTPUT_ACTIVE | LED0_FLAGS) < 0) {
+		return;
+	}
+	gpio_pin_set(gpio_led, PIN0, 0);
```
yes, removed

FRASTM commented on samples/boards/stm32/pm_lptim/src/main.c:
```diff
@@ -0,0 +1,127 @@
+
+	gpio_bn = device_get_binding(SW);
+
+	if (IS_ENABLED(CONFIG_PM)) {
```
ok, removing the condition

pabigot commented on samples/boards/stm32/pm_lptim/README.rst:
```diff
@@ -0,0 +1,46 @@
+   :compact:
+
+After flashing, the LED is on, waiting for entering the low Power mode.
+After WAIT_TIME short delay, the kernel is entering the low power mode (stm32 stop mode),
```
Please don't make the user search the code for what `WAIT_TIME` and `SLEEP_TIME` mean.  Write "a half second delay" and "for four seconds".  Nobody's likely to modify the code to other durations.

FRASTM replied:
Not really sure it is worth to duplicate timing values from code to text, this is just a readme  'FYI'.

pabigot replied:
It really is worth it.  The new user assessing whether something is working will be looking at the README to see what the application is supposed to do.  Please, do not make them look in multiple places to get an accurate understanding of expected behavior.  It will needlessly annoy a potential Zephyr user.

FRASTM replied:
done

FRASTM commented on samples/boards/stm32/pm_lptim/src/main.c:
```diff
@@ -0,0 +1,109 @@
+	"Demo Description\n"	\
+	"Application is demonstrating the LPTIMER use as systick.\n"\
+	"The system goes to sleep mode on k_sleep() and wakes-up\n" \
+	"(red led on) by the lptim timeout expiration\n" \
```
ok

FRASTM commented:
adding control on the regulator to save power

FRASTM commented:
rebase now that #32128 is merged

pabigot commented on samples/boards/stm32/pm_lptim/src/main.c:
```diff
@@ -0,0 +1,109 @@
+/* The devicetree node identifier for the "led0" alias. */
+#define LED0_NODE DT_ALIAS(led0)
+/* GPIO for the Led */
+#define LED0	DT_GPIO_LABEL(LED0_NODE, gpios)
```
It would be easier to understand the code if this were `LED0_LABEL` and the pin were `LED0`.  Then

    gpio_pin_set(gpio_led, LED0, 1);

would be self-explanatory, while:

    gpio_pin_set(gpio_led, PIN0, 1);

is not.

pabigot commented on samples/boards/stm32/pm_lptim/README.rst:
```diff
@@ -0,0 +1,47 @@
+for four seconds. The led is off during this period.
+After four seconds, the timeout wake the system and turn the led on again.
+When pressing the user button, the system is also able to exit from low power mode,
+(led on) because the button is configured in interrupt mode on its EXTI line.
```
It's unclear from this description and running this exactly how this works.

It appears that the cycle is "turn LED on and busy-wait 0.5 sec, then turn LED off and sleep 4 seconds".  Pressing the button at any time turns LED on, and it gets turned off at the end of the next (or current) busy-wait.

So pressing the button does nothing but affect the LED.  I guess.  I suppose we'd have to measure power to see whether there's an effect if the button is pressed 3.5 seconds into the sleep cycle.

abelNoelleC replied:
I can confirm that the expected behaviour you described, indeed is the case. During my testing (on stm32l1_disco) I've observed that the LED will turn on again after pressing the button, but the true link between the LED and the sleep state seems to be missing. 
I've observed that after the button press it does seem to wake up the SoC and kernel, but after that event the LED can still be on even though the SoC is (supposedly) in sleep mode again. In debug I can see that it went through the SoCs power.c file, and the steps for the sleep mode have been taken, but I haven't confirmed whether it actually is in the sleep mode with a power consumption measurement yet. 

I think it would make more sense to see if it's possible to link the current power_state to the LED's output

pabigot reviewed (approved):
I have some non-blocking comments about how the implementation and description might be more clear, but assuming power management were enabled I expect this would work.  I haven't actually tested power measurements to verify it.

pabigot commented:
@noelleclement is this relevant to your use cases?

abelNoelleC commented:
Thanks @pabigot ! A little bit, since it can give me some inspiration for the sample application for the RTC timer. However I'm of the opinion that we should probably have a(n additional) different sample application, since not all boards have on-board LEDs. 

I'll take a more in-depth look tomorrow, and see if I have some specific comments.

abelNoelleC commented on samples/boards/stm32/pm_lptim/README.rst:
```diff
@@ -0,0 +1,47 @@
+.. _Power-Management-LP-Timer:
+
+Power Management LP Timer
+#########################
```
Since this sample is named after the LPTIM, it would be useful to explain why this particular timer is required to test the low-power modes.

abelNoelleC commented on samples/boards/stm32/pm_lptim/README.rst:
```diff
@@ -0,0 +1,47 @@
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/boards/stm32/pm_lptim
+   :board: nucleo_l476rg
```
Is there a way to specify in the folder structure, or at least the README that this is (currently) limited to L4x and WBx?

abelNoelleC reviewed (commented):
I would like to avoid that this is seen as a stm32-wide sample, while it isn't (especially since I'm currently working on the solution for non-LPTIM SoCs, or at least L1x :) ). 
Thank you for picking this up again though @FRASTM !

erwango commented:
> I would like to avoid that this is seen as a stm32-wide sample, while it isn't (especially since I'm currently working on the solution for non-LPTIM SoCs, or at least L1x :) ).
> Thank you for picking this up again though @FRASTM !

I don't think the issue is stm32 here but rather lptim. In the end, this sample should be compatible with any low power solution, be it LPTIM or RTC based. So maybe the point is to rename from `pm_lptim` to `power_mgmt` ?

abelNoelleC commented:
> I don't think the issue is stm32 here but rather lptim. In the end, this sample should be compatible with any low power solution, be it LPTIM or RTC based. So maybe the point is to rename from pm_lptim to power_mgmt ?

Sure, but then again, we need to make sure it's scalable to other power_mgmt implementations, so probably something similar to `samples/drivers/counter/alarm` where there's multiple options for the `TIMER` device? 
Even with this, I think it would still be useful to have some additional information in the sample on why the low-power timer is required (over 'regular' system timer), so the sample can truly be a starting point for setting up their own application.

erwango commented:
> > I don't think the issue is stm32 here but rather lptim. In the end, this sample should be compatible with any low power solution, be it LPTIM or RTC based. So maybe the point is to rename from pm_lptim to power_mgmt ?
> 
> Sure, but then again, we need to make sure it's scalable to other power_mgmt implementations, so probably something similar to `samples/drivers/counter/alarm` where there's multiple options for the `TIMER` device?
> Even with this, I think it would still be useful to have some additional information in the sample on why the low-power timer is required (over 'regular' system timer), so the sample can truly be a starting point for setting up their own application.

I don't think a sample should be the place where board configuration is done.
For sample portability (scalablity as you mention I guess), only the minimum configuration should be done. In that case, sample should enable `CONFIG_PM`. Then it is up to arch/soc/boards configurations to ensure that correct timer setting is achieved when Power Management configuration is requested.
To be clear, when PM is enabled, a L1 board should enable RTC clock automatically, and this should not impact sample implementation.
This way we ensure:
- Easier sample portability
- Easier use of board PM in various applications

abelNoelleC commented:
> I don't think a sample should be the place where board configuration is done.

I agree, but I don't think that would be the case with what I proposed. In that sample it's purely a way to access the device, and its API. However, I looked at it again, and I agree that in this sample that is not required, since the timer driver is more in the background, and indeed dependent on the SoC and board configurations, not really 'part' of the application itself. Whereas in the counter example it's demonstrating the counter API as well. 

So yes, let's not make the sample dependent on the timer. However, I'm still of the opinion that it would be a good idea to mention in the README that for the pm to work, a LP-timer is being used (purely to give people some direction).
